lindbergh: squashfs support

udance4ever opened the pull request:
great work on Lindbergh Loader! 🤯

got .squashfs support working for lindbergh.

1. 94GB for 19 roms compress down to 36GB freeing up 58GB!
2. speeds up updating gamelists as it prevents ES from recursively going down thousands of files in `roms/lindburgh/*`

Tested on:
1. 19 Lindbergh roms
2. 7 other systems that support `.squashfs`: `3ds`, `scummvm`, `Daphne`, `ps3`, `singe`, `snes-msu1`, `xbox`

Tovarichtch commented:
I don't think an overlay is the right solution. However, Sega Lindbergh ROMs really need some level of compression to avoir long gamelist loading and overall management.

Darknior commented:
Ah yes really an excellent idea :)
I love Squashfs format, it same time and hard drive space :)

dmanlfc commented on package/batocera/core/batocera-configgen/configgen/configgen/utils/squashfs.py:
why use overlay?
using the /var/run/squashfs/ as rw is not an ideal location for preservation of any rom updates. why here?

udance4ever replied:
I'm open to other ideas & saw this was the most straightforward to implement and get working.

The bottleneck is `lindbergh` doesn't run outside (eg. `/usr/bin/lindbergh`) and doesn't keep its config files in `system/configs/lindbergh` so it requires a the rom directory to be writeable so the generator can copy files into it.

The solution squashfs already provides is the use an overlay  - that's what it was designed for as far as I can tell.

udance4ever commented:
just heard from the dev of lindbergh.

it sounds like we should be able to run the binary out of `/usr/bin` & create a config file in `~/configs` & leave the NVRAM file in `saves`

I have to assume this has already been tried.

Anyone able to share what the sticking issue was & why we copy binaries into the rom directory?

If I am able to get lindbergh to boot without copying any files, I will make a separate PR so squashfs support is a separate task.

Tovarichtch commented:
This?

https://github.com/lindbergh-loader/lindbergh-loader/blob/efe34307e38cd9a9e84680d125fbf9371a72ab6c/src/lindbergh/lindbergh.c#L285

udance4ever commented:
yes - that was the option we were pointed to.

I just found the sram and eprom options buried in the code:

https://github.com/lindbergh-loader/lindbergh-loader/blob/efe34307e38cd9a9e84680d125fbf9371a72ab6c/src/lindbergh/config.c#L1110-L1114

I assume if we put all the shared libraries (including `libGLcore.so.1`) in `/usr/bin/lindbergh` it will work or is there a reason it is conditionally copied in?

the last wrench in the equation is code that deletes a legacy `libsegaapi.so`  - so if this is necessary, then we are back to requiring the squashfs rom be "writable"

dmanlfc commented:
Previously when I added the emulator the option didn't work. We can retry again.

udance4ever commented:
> Previously when I added the emulator the option didn't work. We can retry again.

ok good to know you've tried!

The shared library issue is the bigger issue.

I added `/usr/bin/lindbergh` to `LD_LIBRARY_PATH` and it still throws an error - is it statically linked to `.` by chance?  Are we able to update the build to dynamically link libraries if this is the case?

I propose we move forward with the .squashfs implementation as is as it does work.

I will resolve merge conflicts and retest when I get the green light.

aderumier commented:
Hi, can't we have some kind of writetable overlay like wsquashfs for windows ?  (maybe made it generic, it could help for mugen too for example)

Darknior commented:
> Hi, can't we have some kind of writetable overlay like wsquashfs for windows ? (maybe made it generic, it could help for mugen too for example)

For sure ... so many files in Linberg games ... it is a pain to transfer or delete by FTP :(
A single file system should help a lot ... but like it is not native, i dont know if Squashfs is possible ...
